From 1ab70c9be4289d049638485d9d48b861f4280809 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:57 -0400
Subject: [powerpc] modules implement R_PPC64_TOCSAVE relocation

Message-id: <1410545655-205645-429-git-send-email-dzickus@redhat.com>
Patchwork-id: 94400
O-Subject: [RHEL7 PATCH 428/626] powerpc: modules implement R_PPC64_TOCSAVE relocation.
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit d247da0a8ebcc4ebb4c766487de6af5df560adac
Author: Rusty Russell <rusty@rustcorp.com.au>
Date:   Tue Mar 18 17:36:28 2014 +1030

    powerpc: modules implement R_PPC64_TOCSAVE relocation.

    Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/uapi/asm/elf.h b/arch/powerpc/include/uapi/asm/elf.h
index 7e39c91..0341109 100644
--- a/arch/powerpc/include/uapi/asm/elf.h
+++ b/arch/powerpc/include/uapi/asm/elf.h
@@ -291,9 +291,12 @@ do {         \
 #define R_PPC64_DTPREL16_HIGHERA 104 /* half16 (sym+add)@dtprel@highera */
 #define R_PPC64_DTPREL16_HIGHEST 105 /* half16 (sym+add)@dtprel@highest */
 #define R_PPC64_DTPREL16_HIGHESTA 106 /* half16 (sym+add)@dtprel@highesta */
+#define R_PPC64_TLSGD  107
+#define R_PPC64_TLSLD  108
+#define R_PPC64_TOCSAVE  109
 
 /* Keep this the last entry.  */
-#define R_PPC64_NUM  107
+#define R_PPC64_NUM  110
 
 /* There's actually a third entry here, but it's unused */
 struct ppc64_opd_entry
diff --git a/arch/powerpc/kernel/module_64.c b/arch/powerpc/kernel/module_64.c
index 7c16b2e..a8694d4 100644
--- a/arch/powerpc/kernel/module_64.c
+++ b/arch/powerpc/kernel/module_64.c
@@ -454,6 +454,14 @@ int apply_relocate_add(Elf64_Shdr *sechdrs,
    *location = value - (unsigned long)location;
    break;
 
+  case R_PPC64_TOCSAVE:
+   /*
+    * Marker reloc indicates we don't have to save r2.
+    * That would only save us one instruction, so ignore
+    * it.
+    */
+   break;
+
   default:
    printk("%s: Unknown ADD relocation: %lu\n",
           me->name,
-- 
1.7.1